### PR TITLE
Feature/Proposal's revision notification service

### PR DESCRIPTION
**Pull request by Nachocss**

### CODE REVIEW

**Comment by Nachocss** [comment]
Let's avoid working on this branch any features that are not strictly related to Proposal's revision notification service. Request a review if you think we should merge it  @mariasoria @juanmaMacGyverCode

**Comment by mariasoria** [comment]
Good morning @ibaiul, how are you?

Could you help us merging this branch, please? Could you have a look at the changes and approve the review?

Thank a million

File: backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java
Comment (manumoranchel) [inline backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java]: As discussed, return 404. Alter Swagger and tests accordingly

File: backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java
Comment (manumoranchel) [inline backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java]: remember not to catch NPEs

File: backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java
Comment (manumoranchel) [inline backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java]: This is not recommended. An API should be agnostic of what client is performing the request. In other words a request can be performed by another application, a web browser, etc. 
Here you are assuming the request is coming from a **browser** and performing a redirect. Such redirect, if needed, should be done from the the **FRONTEND**.

If the request is trying to access an end-point that has been _PERMANTLY_ or _TEMPORARILY_ moved, you can return dedicated _HTTPStatus_ codes. But remember, the client (in huella is the frontend) should perform the redirection

File: backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java
Comment (manumoranchel) [inline backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java:276]: (MINOR) Not necesary to change it.
I can see there is class _com.huellapositiva.domain.model.valueobjects.Roles_ This holds an enum with all the roles handled by the application. Ideally, instead of having "magic strings" we should use the available configured ENUMs.
One of the benefits is when a change is needed, you can change it in one place.

File: backend/src/main/java/com/huellapositiva/application/dto/ProposalLiteDto.java
Comment (manumoranchel) [inline backend/src/main/java/com/huellapositiva/application/dto/ProposalLiteDto.java:19]: Not for this PullRequest
I can see many common fields in the **Proposal*Dto**. Maybe we could use inheritance relationship to avoid duplicate fields?

File: backend/src/main/java/com/huellapositiva/application/exception/ProposalNotPublic.java
Comment (manumoranchel) [inline backend/src/main/java/com/huellapositiva/application/exception/ProposalNotPublic.java]: Not sure what's the agreement in this project since I can see several Exceptions with similar naming.
I usually like to suffix new exceptions with **Exception**, in this case **ProposalNotPublicException** this ease the readiblity and adds context when handling them.
I leave it up to you wether you want to implement it or not

File: backend/src/main/java/com/huellapositiva/domain/actions/FetchPaginatedProposalsAction.java
Comment (manumoranchel) [inline backend/src/main/java/com/huellapositiva/domain/actions/FetchPaginatedProposalsAction.java:17]: I love a good javadoc. Good work

File: backend/src/main/java/com/huellapositiva/domain/actions/RegisterESALAction.java
Comment (manumoranchel) [inline backend/src/main/java/com/huellapositiva/domain/actions/RegisterESALAction.java]: Good work on adding the logs. In my opinion, you should have a log.error right after catching the exception

File: backend/src/main/java/com/huellapositiva/domain/actions/SubmitProposalRevisionAction.java
Comment (manumoranchel) [inline backend/src/main/java/com/huellapositiva/domain/actions/SubmitProposalRevisionAction.java]: what happened here?????? :-)

File: backend/src/main/java/com/huellapositiva/domain/model/entities/Reviser.java
Comment (manumoranchel) [inline backend/src/main/java/com/huellapositiva/domain/model/entities/Reviser.java:8]: can lombok help here?

File: backend/src/main/java/com/huellapositiva/domain/service/ESALContactPersonService.java
Comment (manumoranchel) [inline backend/src/main/java/com/huellapositiva/domain/service/ESALContactPersonService.java]: Shall we define it????

**Review by manumoranchel — COMMENTED** [review]
Looks good overall, good work. I've only commented the code becasue I have no context of the requirements.
I'll leave @ibaiul to have the final decision :-)

File: backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java
Comment (juanmaMacGyverCode) [reply backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java]: Fixed. Test swagger and controller have been modified.

File: backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java
Comment (juanmaMacGyverCode) [reply backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java]: Fixed. Removed the catch and managed null value of the feedback.

File: backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java
Comment (juanmaMacGyverCode) [reply backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java]: Removed the redirection and sent a dedicated HTTP status code and updated tests

File: backend/src/main/java/com/huellapositiva/application/exception/ProposalNotPublic.java
Comment (juanmaMacGyverCode) [reply backend/src/main/java/com/huellapositiva/application/exception/ProposalNotPublic.java]: @carlosble 

We dont know if it is an agreement not to use the word exception at the end of the name the class exception. Do you think we should change it? Many thanks.

File: backend/src/main/java/com/huellapositiva/application/dto/ProposalLiteDto.java
Comment (juanmaMacGyverCode) [reply backend/src/main/java/com/huellapositiva/application/dto/ProposalLiteDto.java:19]: We think that this could be done in a future refactor in develop branch

File: backend/src/main/java/com/huellapositiva/application/exception/ProposalNotPublic.java
Comment (ibaiul) [reply backend/src/main/java/com/huellapositiva/application/exception/ProposalNotPublic.java]: It is a Java convention to suffix with xxxException

File: backend/src/main/java/com/huellapositiva/domain/actions/UploadCurriculumVitaeAction.java
Comment (ibaiul) [inline backend/src/main/java/com/huellapositiva/domain/actions/UploadCurriculumVitaeAction.java]: If cv is null this action does nothing and succeeds. If I want to upload a cv and there is no cv, I would rather fail the request with a 400

File: backend/src/main/java/com/huellapositiva/domain/model/entities/User.java
Comment (ibaiul) [inline backend/src/main/java/com/huellapositiva/domain/model/entities/User.java]: Why not return variable name instead of formatting?

File: backend/src/main/java/com/huellapositiva/domain/model/entities/User.java
Comment (ibaiul) [inline backend/src/main/java/com/huellapositiva/domain/model/entities/User.java]: else if would work too instead of having multiple ifs

File: backend/src/main/java/com/huellapositiva/domain/model/valueobjects/ProposalRevisionEmail.java
Comment (ibaiul) [inline backend/src/main/java/com/huellapositiva/domain/model/valueobjects/ProposalRevisionEmail.java:47]: Logs can help to debug potential issues

File: backend/src/main/java/com/huellapositiva/infrastructure/orm/repository/JpaStatusRepository.java
Comment (ibaiul) [inline backend/src/main/java/com/huellapositiva/infrastructure/orm/repository/JpaStatusRepository.java]: StatusRepository is very ambiguous. What status are we referring to? I would call it JpaProposalStatus otherwise anyone can think it is the status of a volunteer, esal or any other kind of thing.

The same applies to the JpaStatus class

File: backend/src/main/resources/db/migration/V1_0_7__Add_Users_Personal_Information.sql
Comment (ibaiul) [inline backend/src/main/resources/db/migration/V1_0_7__Add_Users_Personal_Information.sql:8]: This is fine for now but once we are live we cannot remove a column this way. It would not allow us to rollback the deployment and would cause complete data corruption. I can explain in a future meeting if interested.

File: backend/src/main/resources/db/migration/V1_0_7__Add_Users_Personal_Information.sql
Comment (ibaiul) [inline backend/src/main/resources/db/migration/V1_0_7__Add_Users_Personal_Information.sql]: If this column is mandatory we need to set a value (default or based on the previous "published" column) to all the existing proposals in the database or the data can get corrupted. It is not that important now since we can delete the DB schema, but it won't be possible in the future once we are live

File: backend/src/main/java/com/huellapositiva/domain/actions/SubmitProposalRevisionAction.java
Comment (mariasoria) [reply backend/src/main/java/com/huellapositiva/domain/actions/SubmitProposalRevisionAction.java]: Ooops, we forgot to add the javadoc.
We've just done it ;)

File: backend/src/main/java/com/huellapositiva/domain/service/ESALContactPersonService.java
Comment (mariasoria) [reply backend/src/main/java/com/huellapositiva/domain/service/ESALContactPersonService.java]: Yes! DONE ;)

File: backend/src/main/resources/db/migration/V1_0_7__Add_Users_Personal_Information.sql
Comment (manumoranchel) [reply backend/src/main/resources/db/migration/V1_0_7__Add_Users_Personal_Information.sql:8]: I understand the reasoning not to drop columns but I would love to hear your experience in rolling back databases to a previous state. It was a challenge in the previous projects I worked and I will always be interested in success stories :-)

File: backend/src/main/java/com/huellapositiva/domain/actions/UploadCurriculumVitaeAction.java
Comment (mariasoria) [reply backend/src/main/java/com/huellapositiva/domain/actions/UploadCurriculumVitaeAction.java]: Changed. Now in this action file we are checking if the file comes empty or if it is not a file attached. In that case, we fail the request with a 400 error.

File: backend/src/main/java/com/huellapositiva/application/exception/ProposalNotPublic.java
Comment (mariasoria) [reply backend/src/main/java/com/huellapositiva/application/exception/ProposalNotPublic.java]: We have just renamed all those exceptions, adding the suffix.

File: backend/src/main/java/com/huellapositiva/domain/model/entities/User.java
Comment (mariasoria) [reply backend/src/main/java/com/huellapositiva/domain/model/entities/User.java]: Changed ;)

File: backend/src/main/java/com/huellapositiva/domain/model/entities/User.java
Comment (mariasoria) [reply backend/src/main/java/com/huellapositiva/domain/model/entities/User.java]: Done!

File: backend/src/main/java/com/huellapositiva/infrastructure/orm/repository/JpaStatusRepository.java
Comment (mariasoria) [reply backend/src/main/java/com/huellapositiva/infrastructure/orm/repository/JpaStatusRepository.java]: Yes, we agree. Changed applied.

File: backend/src/main/resources/db/migration/V1_0_7__Add_Users_Personal_Information.sql
Comment (mariasoria) [reply backend/src/main/resources/db/migration/V1_0_7__Add_Users_Personal_Information.sql:8]: Yes, we would like you to explain us this further, please 🙏

File: backend/src/main/resources/db/migration/V1_0_7__Add_Users_Personal_Information.sql
Comment (mariasoria) [reply backend/src/main/resources/db/migration/V1_0_7__Add_Users_Personal_Information.sql]: We have set the status by default to be REVIEW_PENDING.

File: backend/src/main/java/com/huellapositiva/domain/model/entities/Reviser.java
Comment (mariasoria) [reply backend/src/main/java/com/huellapositiva/domain/model/entities/Reviser.java:8]: Hi Manu, what do you mean? 🎵

Comment (manumoranchel) [reply backend/src/main/java/com/huellapositiva/domain/model/entities/Reviser.java:8]: From a code perspective, the code looks good as it is. I have no argument against it. However Lombok can support with certain common operations like the constructors.
If you want to play with it, I'd look at @NoArgsConstructor, @AllArgsConstructor annotations and see if you can support on them to make the code look nicer.

File: backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java
Comment (ibaiul) [inline backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java]: If the status cannot be set by the user, I would remove the status field from the DTO class (request specification)

File: backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java
Comment (ibaiul) [inline backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java]: Setting the default proposal status seems to be responsibility of the register proposal action rather than the controller

File: backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java
Comment (mariasoria) [reply backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java]: Hi Ibai, you are right. The status is never set by the user. But, depending on the person who creates the proposal, the status is different.
If the “Reviser” creates it, then, the status is directly “PUBLISHED”, if it is the “ContactPerson”, the status is “REVIEW_PENDING” instead.

So, we have different options.

1st option:
If we remove that attribute from the dto object class, still we need to set that value before saving that info into the DB., which can be done in the action class. But, in order to do that (as goes first to the parser to create the object “proposal” which will be the one we will save in the DB), then, we’ll need to add that argument to the parser, which means, sending 3 parameters to it.

2nd option:
If we don’t remove it, even though is actually not part of a dto, we don’t need to do all that from the 1st option.

What do you think? Better remove it from the dto (knowing that that will mean passing more parameters to the parser) or leave it as it is (knowing, that is not actually part of the dto)??

Thank youuu

PS: We will use that “STATUS” soon in other similar situations when the proposal needs to be edited, for example. But that part is not implemented yet.

File: backend/src/main/java/com/huellapositiva/domain/model/entities/Reviser.java
Comment (mariasoria) [reply backend/src/main/java/com/huellapositiva/domain/model/entities/Reviser.java:8]: Hi Manu! Thank you for your comment. 
We had a look at it, and it seems we can't use those annotations, because those constructors extend those in USER.java. Or do you think we can still use them?? Thanks for your help :)

File: backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java
Comment (ibaiul) [reply backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java]: The request DTO should reflect the actual contract of the API, so you should not have an status field and therefore should not deserialise it with the object mapper. I guess that is aligned with your first option.
The internal implementation is up to you, the following are just available options for the action's public contract with the controller.
1- you can define two arguments in the execute method of the action, ProposalRequestDto dto, ProposalStatus status
2- you can create two public methods in the action executeByVolunteer(ProposalRequestDto) and executeByAdmin(ProposalRequestDto dto)
3- you can create two actions that extend an abstract action or implement an abstract action interface

File: backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java
Comment (mariasoria) [reply backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java]: OK, yes, you are right. We have just changed it. Thanks!

File: backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java
Comment (mariasoria) [reply backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java]: Hi Ibai, we think we've got it.
- We removed from the DTO the status (which makes sense). 
- Renamed the "execute" methods, to "executeByContactPerson" and "executeByAdmin".
- Set the status in the action class right after parsing the DTO into a Proposal object. (Not sure if it is the best option) Could you confirm?

Thank you

File: backend/src/main/java/com/huellapositiva/domain/model/entities/Reviser.java
Comment (manumoranchel) [reply backend/src/main/java/com/huellapositiva/domain/model/entities/Reviser.java:8]: I think you are right. Looking at the documentation in: https://projectlombok.org/features/constructor One can see the different configurations that can be added to the mentioned annotations.
Some lombok annotations allow the flag `callSuper = true` However this feature seems not available to build the Object :-)
It is good as it is. Today we've lean something new.

**Review by ibaiul — DISMISSED** [review]
Next time we should ensure we cover %80 of the new code written.

**Review by ibaiul — APPROVED** [review]

**Comment by mariasoria** [comment]
We have merged this branch into develop because the SonarCloud analysis is checking only develop branch, which means that is not even considering the changes done in this branch. After merging, we'll check the analysis again, and will try to solve the issues.

**Pull request merged**
